File: backend/internal/restaurant/deliveryfee.go
package restaurant

import "math"

// Distance/time-based delivery-fee pricing. PURE LOGIC — no DB, no network — so it
// is unit-tested in isolation; the service loads DeliveryFeeConfig from the DB
// (restaurant_delivery_config, admin-adjustable) and feeds the coordinates.
//
//	Fee = round((base + extra-distance + extra-time) × demand_multiplier)
//	      + night_fee + weather_fee + handling_fee − promo_discount, clamped to
//	      [min_fee, max_fee].
//
// Money is integer kobo throughout.

// DeliveryFeeConfig mirrors restaurant_delivery_config (every field admin-editable).
type DeliveryFeeConfig struct {
	BaseFeeKobo       int64   `json:"base_fee_kobo"`
	FreeDistanceKm    float64 `json:"free_distance_km"`
	PerKmKobo         int64   `json:"per_km_kobo"`
	FreeMinutes       float64 `json:"free_minutes"`
	PerMinuteKobo     int64   `json:"per_minute_kobo"`
	DemandMultiplier  float64 `json:"demand_multiplier"`
	NightFeeKobo      int64   `json:"night_fee_kobo"`
	NightStartHour    int     `json:"night_start_hour"`
	NightEndHour      int     `json:"night_end_hour"`
	WeatherFeeKobo    int64   `json:"weather_fee_kobo"`
	HandlingFeeKobo   int64   `json:"handling_fee_kobo"`
	PromoDiscountKobo int64   `json:"promo_discount_kobo"`
	AvgSpeedKmph      float64 `json:"avg_speed_kmph"`
	RoadFactor        float64 `json:"road_factor"`
	MinFeeKobo        int64   `json:"min_fee_kobo"`
	MaxFeeKobo        int64   `json:"max_fee_kobo"` // 0 = no cap
}

// DefaultDeliveryFeeConfig matches the seeded global defaults (the §formula).
func DefaultDeliveryFeeConfig() DeliveryFeeConfig {
	return DeliveryFeeConfig{
		BaseFeeKobo: 80000, FreeDistanceKm: 2, PerKmKobo: 15000,
		FreeMinutes: 25, PerMinuteKobo: 5000, DemandMultiplier: 1.0,
		NightFeeKobo: 0, NightStartHour: 22, NightEndHour: 5,
		WeatherFeeKobo: 0, HandlingFeeKobo: 0, PromoDiscountKobo: 0,
		AvgSpeedKmph: 20, RoadFactor: 1.3, MinFeeKobo: 80000, MaxFeeKobo: 0,
	}
}

// DeliveryFeeBreakdown is the transparent line-by-line result (persisted on the
// order + shown to the buyer).
type DeliveryFeeBreakdown struct {
	DistanceKm   float64 `json:"distance_km"`   // road distance used (km)
	EtaMinutes   float64 `json:"eta_minutes"`   // estimated delivery time (min)
	BaseKobo     int64   `json:"base_kobo"`
	DistanceKobo int64   `json:"distance_kobo"` // extra-distance fee (after free km)
	TimeKobo     int64   `json:"time_kobo"`     // extra-time fee (after free min)
	SurgeKobo    int64   `json:"surge_kobo"`    // extra added by demand multiplier
	NightKobo    int64   `json:"night_kobo"`
	WeatherKobo  int64   `json:"weather_kobo"`
	HandlingKobo int64   `json:"handling_kobo"`
	PromoKobo    int64   `json:"promo_kobo"` // discount applied (subtracted)
	TotalKobo    int64   `json:"total_kobo"`
}

// HaversineKm is the great-circle (straight-line) distance between two lat/lng
// points in kilometres.
func HaversineKm(lat1, lng1, lat2, lng2 float64) float64 {
	const earthKm = 6371.0
	rad := func(d float64) float64 { return d * math.Pi / 180 }
	dLat := rad(lat2 - lat1)
	dLng := rad(lng2 - lng1)
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(rad(lat1))*math.Cos(rad(lat2))*math.Sin(dLng/2)*math.Sin(dLng/2)
	return earthKm * 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
}

// EstimateEtaMinutes estimates delivery time from a road distance + average speed.
func EstimateEtaMinutes(roadKm float64, cfg DeliveryFeeConfig) float64 {
	if cfg.AvgSpeedKmph <= 0 {
		return 0
	}
	return roadKm / cfg.AvgSpeedKmph * 60.0
}

func roundKobo(v float64) int64 {
	if v < 0 {
		return -int64(math.Round(-v))
	}
	return int64(math.Round(v))
}

// ComputeDeliveryFee applies the configurable formula to a STRAIGHT-LINE distance
// (the road factor + ETA are derived inside). Use this when only a haversine
// distance is available (no routing provider). `night`/`weather` toggle the
// respective surcharges. Pure + deterministic.
func ComputeDeliveryFee(straightLineKm float64, night, weather bool, cfg DeliveryFeeConfig) DeliveryFeeBreakdown {
	if straightLineKm < 0 {
		straightLineKm = 0
	}
	roadFactor := cfg.RoadFactor
	if roadFactor < 1 {
		roadFactor = 1
	}
	roadKm := straightLineKm * roadFactor
	eta := EstimateEtaMinutes(roadKm, cfg)
	return feeFromRoute(roadKm, eta, night, weather, cfg)
}

// ComputeDeliveryFeeFromRoute applies the same configurable formula to an ACTUAL
// driving distance (km) + ETA (minutes) from a routing provider (e.g. Google
// Distance Matrix). The road factor and the speed-based ETA estimate are NOT
// applied — the provider already returns real road distance + time. Pure +
// deterministic.
func ComputeDeliveryFeeFromRoute(roadKm, etaMinutes float64, night, weather bool, cfg DeliveryFeeConfig) DeliveryFeeBreakdown {
	if roadKm < 0 {
		roadKm = 0
	}
	if etaMinutes < 0 {
		etaMinutes = 0
	}
	return feeFromRoute(roadKm, etaMinutes, night, weather, cfg)
}

// feeFromRoute is the shared core: given a road distance (km) + ETA (min) it
// applies the distance/time tiers, demand multiplier, surcharges, promo and clamps.
func feeFromRoute(roadKm, eta float64, night, weather bool, cfg DeliveryFeeConfig) DeliveryFeeBreakdown {
	extraKm := roadKm - cfg.FreeDistanceKm
	if extraKm < 0 {
		extraKm = 0
	}
	distanceKobo := roundKobo(float64(cfg.PerKmKobo) * extraKm)

	extraMin := eta - cfg.FreeMinutes
	if extraMin < 0 {
		extraMin = 0
	}
	timeKobo := roundKobo(float64(cfg.PerMinuteKobo) * extraMin)

	core := cfg.BaseFeeKobo + distanceKobo + timeKobo
	mult := cfg.DemandMultiplier
	if mult <= 0 {
		mult = 1
	}
	afterDemand := roundKobo(float64(core) * mult)
	surge := afterDemand - core

	nightKobo := int64(0)
	if night {
		nightKobo = cfg.NightFeeKobo
	}
	weatherKobo := int64(0)
	if weather {
		weatherKobo = cfg.WeatherFeeKobo
	}

	total := afterDemand + nightKobo + weatherKobo + cfg.HandlingFeeKobo - cfg.PromoDiscountKobo

	// Clamp: floor at min (and never below 0), cap at max when set.
	if total < cfg.MinFeeKobo {
		total = cfg.MinFeeKobo
	}
	if cfg.MaxFeeKobo > 0 && total > cfg.MaxFeeKobo {
		total = cfg.MaxFeeKobo
	}
	if total < 0 {
		total = 0
	}

	return DeliveryFeeBreakdown{
		DistanceKm:   math.Round(roadKm*100) / 100,
		EtaMinutes:   math.Round(eta*10) / 10,
		BaseKobo:     cfg.BaseFeeKobo,
		DistanceKobo: distanceKobo,
		TimeKobo:     timeKobo,
		SurgeKobo:    surge,
		NightKobo:    nightKobo,
		WeatherKobo:  weatherKobo,
		HandlingKobo: cfg.HandlingFeeKobo,
		PromoKobo:    cfg.PromoDiscountKobo,
		TotalKobo:    total,
	}
}

// IsNightAt reports whether hour (0-23, delivery-locale) falls in the night window
// (handles windows that wrap past midnight, e.g. 22→5).
func IsNightAt(hour int, cfg DeliveryFeeConfig) bool {
	s, e := cfg.NightStartHour, cfg.NightEndHour
	if s == e {
		return false
	}
	if s < e {
		return hour >= s && hour < e
	}
	return hour >= s || hour < e // wraps midnight
}

File: mobile-app/reactnative/src/features/food/deliveryFee.ts
// ── Restaurant & Delivery — Delivery-fee MOCK calculator ─────────────────────
// Pure, deterministic mirror of backend/internal/restaurant/deliveryfee.go so the
// food checkout can quote a distance/time-based delivery fee fully offline. The
// SERVER stays authoritative on placeOrder — this only powers the pre-payment
// estimate (and the mock when EXPO_PUBLIC_FOOD_USE_MOCK !== 'false').
//
//   Fee = round((base + extra-distance + extra-time) × demand)
//         + night + weather + handling − promo, clamped to [min, max].
//
// Money is integer kobo throughout (never floats for the result lines).

import type { LatLng } from './types';

/** Mirrors restaurant.DeliveryFeeConfig (every field admin-editable server-side). */
export interface DeliveryFeeConfig {
  baseFeeKobo: number;
  freeDistanceKm: number;
  perKmKobo: number;
  freeMinutes: number;
  perMinuteKobo: number;
  demandMultiplier: number;
  nightFeeKobo: number;
  weatherFeeKobo: number;
  handlingFeeKobo: number;
  promoDiscountKobo: number;
  avgSpeedKmph: number;
  roadFactor: number;
  minFeeKobo: number;
  /** 0 = no cap. */
  maxFeeKobo: number;
}

/** Matches restaurant.DefaultDeliveryFeeConfig() — the seeded global defaults. */
export const DEFAULT_DELIVERY_CONFIG: DeliveryFeeConfig = {
  baseFeeKobo: 80000, // ₦800 base
  freeDistanceKm: 2,
  perKmKobo: 15000, // ₦150 / km after 2 km
  freeMinutes: 25,
  perMinuteKobo: 5000, // ₦50 / min after 25 min
  demandMultiplier: 1.0,
  nightFeeKobo: 0,
  weatherFeeKobo: 0,
  handlingFeeKobo: 0,
  promoDiscountKobo: 0,
  avgSpeedKmph: 20,
  roadFactor: 1.3,
  minFeeKobo: 80000,
  maxFeeKobo: 0,
};

/** Transparent line-by-line result — same shape as the backend breakdown JSON. */
export interface DeliveryFeeBreakdown {
  distance_km: number;
  eta_minutes: number;
  base_kobo: number;
  distance_kobo: number;
  time_kobo: number;
  surge_kobo: number;
  night_kobo: number;
  weather_kobo: number;
  handling_kobo: number;
  promo_kobo: number;
  total_kobo: number;
}

/** Full quote envelope — same shape the /delivery-quote endpoint returns. */
export interface DeliveryQuote {
  delivery_fee_kobo: number;
  flat_fallback: boolean;
  breakdown?: DeliveryFeeBreakdown;
}

/** How the delivery fee should be shown, and what to add to the estimate. */
export interface DeliveryFeeView {
  /** Kobo to include in the estimated total. 0 only when `known` is false. */
  feeKobo: number;
  /** True when this is a real number the SERVER returned. */
  known: boolean;
  /** True when the number is not distance-based, so it may move at checkout. */
  estimated: boolean;
}

/**
 * Turn a quote into what the summary should display.
 *
 * The rule this encodes: a quote is a SERVER price and is always used.
 * `flat_fallback` means "not distance-based" — it does NOT mean "discard".
 *
 * Checkout used to require `!flat_fallback` before trusting the quote and
 * otherwise fell back to `restaurant.deliveryFeeKobo`, a field with no database
 * column and no DTO behind it, so the fallback was always 0. The server returns
 * flat_fallback whenever the restaurant has no coordinates — 653 of 697 in the
 * dev DB — so the normal path threw away a real ₦500 and rendered ₦0, while
 * PlaceOrder went on charging the server-computed fee.
 *
 * With no quote at all the fee is UNKNOWN, not free: callers must render that
 * as a placeholder, never as ₦0.
 */
export function resolveDeliveryFee(quote?: DeliveryQuote | null): DeliveryFeeView {
  if (!quote || typeof quote.delivery_fee_kobo !== 'number' || !Number.isFinite(quote.delivery_fee_kobo)) {
    return { feeKobo: 0, known: false, estimated: true };
  }
  return {
    feeKobo: Math.trunc(quote.delivery_fee_kobo),
    known: true,
    estimated: Boolean(quote.flat_fallback),
  };
}

/**
 * Great-circle (straight-line) distance between two lat/lng points in kilometres.
 * Mirrors restaurant.HaversineKm.
 */
export function haversineKm(a: LatLng, b: LatLng): number {
  const earthKm = 6371.0;
  const rad = (d: number) => (d * Math.PI) / 180;
  const dLat = rad(b.lat - a.lat);
  const dLng = rad(b.lng - a.lng);
  const h =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(rad(a.lat)) * Math.cos(rad(b.lat)) * Math.sin(dLng / 2) * Math.sin(dLng / 2);
  return earthKm * 2 * Math.atan2(Math.sqrt(h), Math.sqrt(1 - h));
}

/** Round to integer kobo, matching Go's roundKobo (round-half-away-from-zero). */
function roundKobo(v: number): number {
  if (v < 0) return -Math.round(-v);
  return Math.round(v);
}

/**
 * Apply the configurable formula to the restaurant→delivery straight-line distance
 * (road factor + ETA derived inside). Pure + deterministic; mirrors
 * restaurant.ComputeDeliveryFee. Returns the line-by-line breakdown.
 */
export function computeDeliveryFeeMock(
  restaurantLoc: LatLng,
  deliveryLoc: LatLng,
  cfg: DeliveryFeeConfig = DEFAULT_DELIVERY_CONFIG,
  flags: { night?: boolean; weather?: boolean } = {},
): DeliveryFeeBreakdown {
  let straightLineKm = haversineKm(restaurantLoc, deliveryLoc);
  if (straightLineKm < 0) straightLineKm = 0;

  let roadFactor = cfg.roadFactor;
  if (roadFactor < 1) roadFactor = 1;
  const roadKm = straightLineKm * roadFactor;
  const eta = cfg.avgSpeedKmph > 0 ? (roadKm / cfg.avgSpeedKmph) * 60.0 : 0;

  const extraKm = Math.max(0, roadKm - cfg.freeDistanceKm);
  const distanceKobo = roundKobo(cfg.perKmKobo * extraKm);

  const extraMin = Math.max(0, eta - cfg.freeMinutes);
  const timeKobo = roundKobo(cfg.perMinuteKobo * extraMin);

  const core = cfg.baseFeeKobo + distanceKobo + timeKobo;
  let mult = cfg.demandMultiplier;
  if (mult <= 0) mult = 1;
  const afterDemand = roundKobo(core * mult);
  const surge = afterDemand - core;

  const nightKobo = flags.night ? cfg.nightFeeKobo : 0;
  const weatherKobo = flags.weather ? cfg.weatherFeeKobo : 0;

  let total = afterDemand + nightKobo + weatherKobo + cfg.handlingFeeKobo - cfg.promoDiscountKobo;

  // Clamp: floor at min (and never below 0), cap at max when set.
  if (total < cfg.minFeeKobo) total = cfg.minFeeKobo;
  if (cfg.maxFeeKobo > 0 && total > cfg.maxFeeKobo) total = cfg.maxFeeKobo;
  if (total < 0) total = 0;

  return {
    distance_km: Math.round(roadKm * 100) / 100,
    eta_minutes: Math.round(eta * 10) / 10,
    base_kobo: cfg.baseFeeKobo,
    distance_kobo: distanceKobo,
    time_kobo: timeKobo,
    surge_kobo: surge,
    night_kobo: nightKobo,
    weather_kobo: weatherKobo,
    handling_kobo: cfg.handlingFeeKobo,
    promo_kobo: cfg.promoDiscountKobo,
    total_kobo: total,
  };
}
